src/input/MeshAttributes.cpp

```diff
@@ -38,250 +35,123 @@ void MeshAttributes::readXmlFile(const char* xmlFilename) {
-
-       child2 = child->FirstChildElement("Height");
-       sval = child2->Attribute("x");
-       mycube.CubeHeight[0] =  atof(sval.c_str());
-       sval = child2->Attribute("y");
-       mycube.CubeHeight[1] =  atof(sval.c_str());
-       sval = child2->Attribute("z");
-       mycube.CubeHeight[2] =  atof(sval.c_str());
-
-       child2 = child->FirstChildElement("Depth");
-       sval = child2->Attribute("x");
-       mycube.CubeDepth[0] =  atof(sval.c_str());
-       sval = child2->Attribute("y");
-       mycube.CubeDepth[1] =  atof(sval.c_str());
-       sval = child2->Attribute("z");
-       mycube.CubeDepth[2] =  atof(sval.c_str());
+       Cube mycube;
+       mycube.CubeMSize =  std::atof(child->Attribute("value"));
+       for (std::string::size_type i = 0; i < xyz.size(); i++) {
+           retrieve_attribute_and_update_class_member(child, "Center", &xyz[i], &mycube.CubeCenter[i]);
+           retrieve_attribute_and_update_class_member(child, "Width", &xyz[i], &mycube.CubeWidth[i]);
+           retrieve_attribute_and_update_class_member(child, "Height", &xyz[i], &mycube.CubeHeight[i]);
+           retrieve_attribute_and_update_class_member(child, "Depth", &xyz[i], &mycube.CubeDepth[i]);
+       }
        lCube.push_back(mycube);
     }
 }
 
 
 void MeshAttributes::set_SurfaceMeshingAttributes() {
-    std::string sval;
-    XMLElement* pRoot;
-    pRoot = doc.FirstChildElement("SurfaceMeshing");
-    if(pRoot) {
-       sval = pRoot->Attribute("SmoothingLevel");
-       surfaceSmoothingLevel  =  atoi(sval.c_str());
-       sval = pRoot->Attribute("Snap");
-       surfaceSnap  =  atoi(sval.c_str());
-       sval = pRoot->Attribute("SmoothingType");
-       if (sval.compare("Laplacian")==0) {
-          surfaceSmoothingType=0;
-       } else if (sval.compare("Gradient")==0) {
-          surfaceSmoothingType=1;
-       } else {
-          logError() << "Unrecognised surfaceSmoothingType (Laplacian or Gradient)" << sval;
-       }
-       sval = pRoot->Attribute("DiscreteAngle");
-       surfaceFaceRotationLimit  =  atof(sval.c_str());
-       logInfo(PMU_rank()) << "surface smoothing option: surfaceSmoothingLevel surfaceSmoothingType surfaceFaceRotationLimit Snap"<<
-         surfaceSmoothingLevel<<" "<< surfaceSmoothingType<<" "<< surfaceFaceRotationLimit <<" "<<surfaceSnap;
+    retrieve_attribute_and_update_class_member(&doc, "SurfaceMeshing", "DiscreteAngle", &surfaceFaceRotationLimit);
+    retrieve_attribute_and_update_class_member(&doc, "SurfaceMeshing", "SmoothingLevel", &surfaceSmoothingLevel);
+    retrieve_attribute_and_update_class_member(&doc, "SurfaceMeshing", "Snap", &surfaceSnap);
+    if (auto pRoot = doc.FirstChildElement("SurfaceMeshing")) {
+        std::string sSmoothingType = pRoot->Attribute("SmoothingType");
+        if (sSmoothingType.compare("Laplacian")==0) {
+           surfaceSmoothingType=0;
+        } else if (sSmoothingType.compare("Gradient")==0) {
+           surfaceSmoothingType=1;
```

Comment:
You could use a class enum for surfaceSmoothingType.

Reply:
How would you do that?
https://stackoverflow.com/questions/43510615/enum-class-of-type-string-in-c

Reply:
Have somewhere in the Header:
```
enum class MeshAttributes::SmoothingType {
  Laplacian,
  Gradient
};
```
And then 
```
if (sSmoothingType.compare("Laplacian")==0) {
           surfaceSmoothingType=SmoothingType::Laplacian;
```

Reply:
ok thx.

Reply:
Hi, actually I don't really understand how to use that.
surfaceSmoothingType has to be an int not an enum, e.g., if I do 

```
SmoothingType surfaceSmoothingType = SmoothingType::Laplacian;
```
Then I get:
```
/hppfs/work/pr45fi/di73yeq4/myLibs/PUMGen/src/input/SimModSuite.h(158): error: argument of type "MeshAttributes::SmoothingType" is incompatible with parameter of type "int"
                     VolumeMesher_setSmoothType(volumeMesher, MeshAtt.volumeSmoothingType);
```

but then if I try:

```
using namespace tinyxml2;
class MeshAttributes {
  public:
    enum class SmoothingType: int {
      Laplacian = 0,
      Gradient = 1
    };
    int surfaceSmoothingLevel=2;
    int surfaceSmoothingType=1;
```
I get:
```
/hppfs/work/pr45fi/di73yeq4/myLibs/PUMGen/src/input/MeshAttributes.cpp(78): error: a value of type "MeshAttributes::SmoothingType" cannot be assigned to an entity of type "int"
             volumeSmoothingType = SmoothingType::Laplacian;
                                 ^
```

Reply:
Try to change `SimModSuite.h(158)` to `VolumeMesher_setSmoothType(volumeMesher, static_cast<int>(MeshAtt.volumeSmoothingType));`
Why so complicated you might ask? Using the Simmetrix conventions (Laplacian=0, Gradient=1) everywhere in the code, makes it hard to maintain and error prone. If you are programming the xml reader, you always have to keep in mind, what their convention was. With the strongly typed enum, we just convert to int when we call the Simmetrix library acoording to their conventions.

Reply:
ok!